[WebCodecs] AudioDecoder follow-up after 267403@main

philn opened the pull request:
#### 1fb4cdd856dd56a4c6b7973ed2a798f8d2e3a793
<pre>
[WebCodecs] AudioDecoder follow-up after 267403@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=260958">https://bugs.webkit.org/show_bug.cgi?id=260958</a>

Reviewed by Youenn Fablet.

The AudioDecoder config validation now follows the spec more closely. Some tests still fail, not
catching the exception raised when configuring the decoder with an invalid config, but the same
failures happen in the VideoDecoder, so both should be addressed in a separate patch.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any.worker-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/video-encoder-config.https.any-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/video-encoder-config.https.any.worker-expected.txt:
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
(WebCore::isValidDecoderConfig):
(WebCore::createAudioDecoderConfig):
(WebCore::WebCodecsAudioDecoder::configure):
(WebCore::WebCodecsAudioDecoder::isConfigSupported):
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoderConfig.h:
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoderConfig.idl:
* Source/WebCore/platform/AudioDecoder.cpp:
(WebCore::AudioDecoder::isCodecSupported):

Canonical link: <a href="https://commits.webkit.org/267743@main">https://commits.webkit.org/267743@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/327f89e7f58ad36004c86fe91a02d8234a9c5cb1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17157 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17482 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17975 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18943 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16055 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17349 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17624 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18250 "16 flakes 150 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17360 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17704 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14892 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19759 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14944 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15583 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22273 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15943 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15750 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20091 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16343 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13862 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15497 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4182 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19864 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16174 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -63,9 +63,27 @@ WebCodecsAudioDecoder::~WebCodecsAudioDecoder()
+    if (StringView(config.codec).trim(isASCIIWhitespace<UChar>).isEmpty())
+        return false;
+
+    if (!config.numberOfChannels)
```
sampleRate and numberOfChannels should probably be made required in WebCodecsAudioDecoderConfig.idl instead.

philn replied:
Indeed, the initial implementation is incorrect.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -63,9 +63,27 @@ WebCodecsAudioDecoder::~WebCodecsAudioDecoder()
+        return false;
+
+    if (!*config.numberOfChannels)
+        return false;
```
I do not see this check and some below in https://w3c.github.io/webcodecs/#valid-audiodecoderconfig.
Where is it coming from?

philn replied:
Well the WPT tests check that, and I see blink does too. Haven't checked gecko though.

philn replied:
I'll revert this part and add a FIXME.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
```diff
@@ -63,9 +63,27 @@ WebCodecsAudioDecoder::~WebCodecsAudioDecoder()
+        return false;
+
+    if ((config.codec == "flac"_s || config.codec == "vorbis"_s ) && !config.description)
+        return false;
```
I do not see this check in the spec. I would think configure would close the decoder instead.

philn replied:
That's mandated by the flac and vorbis codec registries. Blink has similar code, but I can remove it. AFAICS there's no WPT test checking this.

philn replied:
Removed.

philn committed:
[WebCodecs] AudioDecoder follow-up after 267403@main
https://bugs.webkit.org/show_bug.cgi?id=260958

Reviewed by Youenn Fablet.

The AudioDecoder config validation now follows the spec more closely. Some tests still fail, not
catching the exception raised when configuring the decoder with an invalid config, but the same
failures happen in the VideoDecoder, so both should be addressed in a separate patch.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/audio-decoder.https.any.worker-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/video-encoder-config.https.any-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/video-encoder-config.https.any.worker-expected.txt:
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp:
(WebCore::isValidDecoderConfig):
(WebCore::createAudioDecoderConfig):
(WebCore::WebCodecsAudioDecoder::configure):
(WebCore::WebCodecsAudioDecoder::isConfigSupported):
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoderConfig.h:
* Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoderConfig.idl:
* Source/WebCore/platform/AudioDecoder.cpp:
(WebCore::AudioDecoder::isCodecSupported):

Canonical link: https://commits.webkit.org/267743@main
